internal/pkg/template/override/rule.go

```diff
@@ -60,7 +62,66 @@ func (r Rule) validate() error {
 }
 
 func (r Rule) parse() (nodeUpserter, error) {
-	return nil, nil
+	pathSegments := strings.SplitN(r.Path, pathSegmentSeparator, 2)
+	subMatches := pathSegmentRegexp.FindStringSubmatch(pathSegments[0])
+	if len(subMatches) == 0 {
+		// This error shouldn't occur given that `validate()` has passed.
+		return nil, fmt.Errorf(`invalid override path segment "%s"`, pathSegments[0])
+	}
+	// https://pkg.go.dev/regexp#Regexp.FindStringSubmatch
+	// Given that path segment is valid (after `validate()`), `subMatches` contains four elements.
+	// subMatches[0] is the whole path segment, e.g. ContainerDefinitions[0].
+	// subMatches[1:] are individual capture groups. There are 3 capture groups.
+	// subMatches[1], the first capture group is "([a-zA-Z0-9_-]+)", i.e. the `key`.
+	// subMatches[2], the second capture group is "[<index>]".
+	// subMatches[3], the third capture group us "<index>".
+	// subMatches[2] and subMatches[3] could also be empty string, depending on whether there is "[<i>]" in path segment.
+	key := subMatches[1]
```

Comment:
nit: in a follow-up PR i think it'd be cool to actually move this info to struct so that we don't need the comments for example:
```go
type pathSegment struct {
   raw string
   
   // details on the segment.
   key string
   ref string
   index string
}

func parsePathSegment(raw string) (pathSegment, error)
```

Reply:
will follow up!